### PR TITLE
[MOD-2078] Add memory limit specification

### PR DESCRIPTION
**Pull request by thundergolfer**
## Describe your changes

- MOD-2078

The following program will set a memory limit of 2 GiB on the container, which will cause OOM failure when run. 

```python
import modal
import random

stub = modal.Stub()

@stub.function(memory=(1024, 2048))
def main():
    ten_gib = 4 * 1024 * 1024 * 1024  # 4 GiB in bytes
    memory_hog = bytearray(ten_gib)
    for i in range(10_000_000):
        memory_hog[i] = random.randrange(128) # mutate
    print("done")
```

<kbd>
image
</kbd>

## Backward/forward compatibility checks

Check these boxes or delete any item (or this section) if not relevant for this PR.

- [x] Client+Server: this change is compatible with old servers
    - old servers will ignore the `memory_mb_max` value.
- [x] Client forward compatibility: this change ensures client can accept data intended for later versions of itself

## Changelog

* Specifying a hard memory limit for a `modal.Function` is now supported. Pass a tuple of `memory=(request, limit)`. Above the `limit`, which is specified in MiB, a Function's container will be OOM killed.

### CODE REVIEW

**Comment by mwaskom** [comment]
I think you want to update the type / comment in `Stub.cls` too — super annoying that we need to manually keep these synced. Not completely certain there's nowhere else...

**Comment by thundergolfer** [comment]
Oh yeh thanks!

**Comment by thundergolfer** [comment]
@prbot approve

File: modal/functions.py [hunk 1]
```diff
@@ -802,6 +802,16 @@ async def _load(self: _Function, resolver: Resolver, existing_object_id: Optiona
...
+            if memory and isinstance(memory, int):
+                memory_mb = memory
+                memory_mb_max = 0  # no limit
+            elif memory and isinstance(memory, tuple):
+                memory_mb, memory_mb_max = memory
+                if memory_mb_max < memory_mb:
+                    raise ValueError(f"Cannot specify a memory limit lower than request: {memory_mb_max} < {memory_mb}")
+            else:
+                memory_mb = 0
+                memory_mb_max = 0
```
Comment (mwaskom) [inline modal/functions.py:814]: Kind of annoying that we need to have this logic duplicated :/

File: modal/cls.py [hunk 2]
```diff
@@ -292,7 +292,19 @@ def with_options(
...
             milli_cpu = int(1000 * cpu) if cpu is not None else None
             gpu_config = parse_gpu_config(gpu)
-            resources = api_pb2.Resources(milli_cpu=milli_cpu, gpu_config=gpu_config, memory_mb=memory)
+            if memory and isinstance(memory, int):
+                memory_mb = memory
+                memory_mb_max = 0  # no limit
+            elif memory and isinstance(memory, tuple):
+                memory_mb, memory_mb_max = memory
+                if memory_mb_max < memory_mb:
+                    raise ValueError(f"Cannot specify a memory limit lower than request: {memory_mb_max} < {memory_mb}")
```
Comment (mwaskom) [inline modal/cls.py:301]: It's a little unclear to me whether we have a policy of always raising `modal.exception` classes from the client or whether that is even a good idea, but I think that usually we would raise an `InvalidError` here? Although `ValueError` is also correct.

**Review by mwaskom — APPROVED** [review]
Good feature; not wild about the amount of code duplication, curious if we need to do some refactoring at some point.

File: modal/functions.py (hunk 1 above)
Comment (thundergolfer) [reply modal/functions.py:814]: Yeh I could have factored it out into a helper 😅 . Rule of threes.

File: modal/cls.py (hunk 2 above)
Comment (thundergolfer) [reply modal/cls.py:301]: Yeh also thought the same. I think `ValueError` is correct, but I'd also want to things to be consistently correct across the codebase.

Comment (mwaskom) [reply modal/cls.py:301]: I'm sure this isn't the only instance; maybe we should just make a ticket and try to standardize prior to v1.0

**Pull request merged**